### PR TITLE
Prepare for concurrent IOVs in Geometry

### PR DESCRIPTION
**Pull request by wddgit**
Change return type of function producing data from shared_ptr to unique_ptr in
many Geometry ESProducers.

### CODE REVIEW

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade

File: Geometry/HGCalCommonData/plugins/FastTimeNumberingInitialization.cc [hunk 1]
```diff
@@ -66,12 +63,9 @@ FastTimeNumberingInitialization::produce(const IdealGeometryRecord& iRecord) {
 #ifdef EDM_ML_DEBUG
   std::cout << "in FastTimeNumberingInitialization::produce" << std::endl;
 #endif
-  if (fastTimeDDDConst_ == nullptr) {
```
Comment (fabiocos) [inline Geometry/HGCalCommonData/plugins/FastTimeNumberingInitialization.cc]: @wddgit for my understanding, why did you remove such protection?

Comment (wddgit) [reply Geometry/HGCalCommonData/plugins/FastTimeNumberingInitialization.cc]: Hi Fabio,

Before the changes in this PR, I think this ESProducer had a bug. It created a FastTimeDDDConstants object on the first call to produce and it saved a raw pointer to that object as member data. It is important that it saved not a shared pointer but a raw point. Then in subsequent calls to the produce functions, the condition in the "if" you are asking about would be false. So it would reuse the object using the raw pointer by constructing an entirely new shared_ptr from the raw pointer and returning it to the Framework. (Note that it only created a new shared_ptr, not a new object!) At the end of an IOV the Framework would reset the shared_ptr. Because the module is only holding a raw pointer, the object would actually be destroyed. So if the produce method was called more than one time, the same object would be deleted multiple times (and double deletes usually cause a seg fault and the process crashes).

There are two solutions I could think of:

Option 1. Delete the data member entirely and do not save a pointer to the object at all. Recreate a new object on every call to produce. Return a unique_ptr and let the Framework destroy the object at the end of the IOV. This is what I did. The member variable fastTimeDDDConst_ no longer exists after this PR so you cannot check whether or not it is null.

Option 2. Change the class member variable to be a shared_ptr so the module keeps the object from being destroyed when the IOV ends. This would add complication to the code. Then I would also have to add protection so that all this would handled properly with multiple concurrent IOVs. I would add a ReusableObjectHolder and make sure the concurrency issues are handled properly (with multiple concurrent IOVs there would have to one object for each IOV being processed at the same time).

I chose option 1 because there didn't seem to be enough motivation to add the extra code complexity to optimize it. Probably this ESProducer's produce function is never called twice, because no has reported a crash here (I don't even know if it is ever called at all). I looked at the constructor and see it is doing some work to build the object, but it does not look like so much work as to be significant. If someone did a performance profile that showed it was significant, then we should optimize and implement option 2.

Maybe too long winded an answer, but that was my thought process.

Regards
David Dagenhart

Comment (Dr15Jones) [reply Geometry/HGCalCommonData/plugins/FastTimeNumberingInitialization.cc]: I endorse option 1.
